Add Generate LODs, Shadow Mesh and Lightmap UV2 options to OBJ mesh import

Calinou opened the pull request:
- Follow-up to https://github.com/godotengine/godot/pull/94097
(can be merged independently).
- Depends on https://github.com/godotengine/godot/pull/94100.

This puts OBJ mesh import on parity with 3D scene import. It's now possible to have the same level of optimization as imported 3D scenes while using the OBJ mesh workflow.

LOD and shadow mesh generation are enabled by default, so that users can benefit from the same performance regardless of how they're importing 3D objects into Godot. Lightmap UV2 generation is disabled by default to mimic the 3D scene import default (and PrimitiveMesh).

`optimize_indices_for_cache()` is now always called on import as well, similar to what the 3D scene import already does.

- See https://github.com/godotengine/godot-proposals/issues/5600.

**Testing project:** [test_obj_import_lod_shadow_mesh_lightmap_uv2.zip](https://github.com/user-attachments/files/16135187/test_obj_import_lod_shadow_mesh_lightmap_uv2.zip)

## Preview

### OBJ mesh import options

![image](https://github.com/godotengine/godot/assets/180032/88db7cc9-7bb5-4aa6-82bd-1fda245818a1)

### Lightmapped OBJ mesh having its LOD level changed

https://github.com/godotengine/godot/assets/180032/0bd84056-7628-4583-9792-3e8a44c9205f

### LOD disabled

![Screenshot_20240709_002424 png webp](https://github.com/godotengine/godot/assets/180032/5956902f-9204-4214-b502-663882cbf81a)

### LOD enabled (Mesh LOD Threshold = 1024)

*This shows the most decimated LODs for comparison purposes.*

![Screenshot_20240709_002431 png webp](https://github.com/godotengine/godot/assets/180032/21c4aa62-6cbe-4dd0-a36a-ef0ccc29e655)

## TODO

- [x] Figure out why it runs much slower than 3D scene import when LOD generation or lightmap UV2 generation is enabled. (Shadow mesh generation also suffers from this problem to a less severe extent.)
  - This does not occur with 3D scene import, even on the same OBJ file switched to use the **OBJ as Scene** import mode.

Calinou committed:
Add Generate LODs, Shadow Mesh and Lightmap UV2 options to OBJ mesh import

This puts OBJ mesh import on parity with 3D scene import. It's now
possible to have the same level of optimization as imported 3D scenes
while using the OBJ mesh workflow.

`optimize_indices_for_cache()` is now always called on import as well,
similar to what the 3D scene import already does.

Calinou commented on editor/import/3d/resource_importer_obj.cpp:
```diff
@@ -445,6 +445,42 @@ static Error _parse_obj(const String &p_path, List<Ref<ImporterMesh>> &r_meshes,
+					}
+				}
+
+				if (p_generate_lods) {
+					// Use normal merge/split angles that match the defaults used for 3D scene importing.
+					mesh->generate_lods(60.0f, 25.0f, {});
+				}
+
+				if (p_generate_shadow_mesh) {
+					mesh->create_shadow_mesh();
+				}
```
Note: if you perform these steps *before* lightmap UV2 generation, the resulting mesh will be broken. It'll display with the wrong transform and no material (pure black color). I have no idea why, and putting up a MRP will likely require exposing additional ImporterMesh methods (such as `create_shadow_mesh()`).

fracteed commented:
Nice work. I would look to see an option to not to try to import .mat files when importing an obj. It is a constant annoyance to see errors in the console that it can't find the corresponding .mat file for an obj. 

Personally, I never use the .mat file, so only export it to stop Godot spitting out the error. Godot will see face groups fine with just the .obj itself.

lyuma commented:
> I would look to see an option to not to try to import .mat files

@fracteed To be clear, you're talking about .mtl files referenced by a `mtllib` statement in an obj

My opinion is that it is data which is directly referenced by the obj file, so it would be wrong not to try to load it.

I don't think we should add an option to disable an error, especially when there are cases where the error is helpful. If you really don't have the .mtl, it's also relatively easy to edit the .obj file and remove the `mtllib` statement.

fracteed commented:
@lyuma yeah, I meant to write mtl. There should at least be an option to turn off the requirement for it for those that just need the mesh information. If I have hundreds of objs in a project, it is not viable to fiddle with editing obj files.

Calinou commented:
> It is a constant annoyance to see errors in the console that it can't find the corresponding .mat file for an obj.

[A PR to address this was just merged](https://github.com/godotengine/godot/pull/94102) :slightly_smiling_face:

akien-mga commented:
I think we should add an import option for OBJ to ignore MTL files (i.e. just skip `mtllib` references). IIRC Godot doesn't really support the old specular mapping they use, so only bits of it can be imported. More often than not people might want to write their own material and not use the one a thirdparty OBJ file may have come with.

zeux commented:
I think the extra processing steps should be done in import_scene or at the end of parse obj, not inside the loop? An .obj file with multiple materials or surface groups will add multiple surfaces to the same mesh, but with this change shadow mesh or LOD or lightmap generation will be called on the entire mesh for every newly added surface. For lightmaps this might be mitigated with a cache, but for LOD/shadow meshes it looks like the code will redundantly reprocess all earlier computed surfaces.

zeux commented:
Also after https://github.com/godotengine/godot/pull/94241 it would make sense to call `ImporterMesh::optimize_indices_for_cache` as well; the improvements in create_shadow_mesh will automatically apply here with no changes.

clayjohn commented:
Since https://github.com/godotengine/godot/pull/94241 has been merged. Could you add `ImporterMesh::optimize_indices_for_cache` as part of this PR? I would love to include this in 4.4

Calinou commented:
Rebased and tested again, it works as expected.

> I think the extra processing steps should be done in import_scene or at the end of parse obj, not inside the loop? An .obj file with multiple materials or surface groups will add multiple surfaces to the same mesh, but with this change shadow mesh or LOD or lightmap generation will be called on the entire mesh for every newly added surface. For lightmaps this might be mitigated with a cache, but for LOD/shadow meshes it looks like the code will redundantly reprocess all earlier computed surfaces.

That was the cause of the performance issue :slightly_smiling_face:

Fixed, thanks for the pointer!

It's now possible to use lightmap unwrap on `octavus.obj` in the testing project with reasonable performance:

![image](https://github.com/user-attachments/assets/96e849b8-62f7-40d9-a23b-4bfabe1c6486)

> Since #94241 has been merged. Could you add `ImporterMesh::optimize_indices_for_cache` as part of this PR? I would love to include this in 4.4

Done.

clayjohn reviewed (approved):
Looks great to me!

fire reviewed (approved):
The proposal makes sense to me and is wanted.

Did not technical review.

Repiteo commented:
Thanks!
